### PR TITLE
new (dry) way to compose where clauses in repos

### PR DESCRIPTION
**Pull request by joshm1**
the motivation is that methods in repos contain a lot of duplicate code.  I'm introducing a syntax that is shorter, (subjectively) as readable , and could eliminate a lot of duplication between methods that have several identical filters.  i tested queries produced by slick and they are the same as before, so no database impact in this case.

the pipe `|>` syntax is a method @andrewconner created that essentially acts like the unix pipe and allows us to compose these `RepoQuery` methods without deeply nesting them like `h(g(f(x)))`

@eishay @stkem @ymatsuda @andrewconner @ahsu1230 @lawzlo @2is10 @raymondng @leogrim

### CODE REVIEW

**Comment by andrewconner** [comment]
How bad is the generated SQL? :smile: 

edit: "i tested queries produced by slick" Oh wow, surprising. Nice.

**Comment by yingjieMiao** [comment]
nice!

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 1]
```diff
@@ -43,6 +44,27 @@ class UriRecommendationRepoImpl @Inject() (
 
   type RepoImpl = UriRecommendationTable
 
+  object QueryBuilder {
```
Comment (2is10) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: `private`?

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 2]
```diff
@@ -63,22 +85,19 @@ class UriRecommendationRepoImpl @Inject() (
   initTable()
 
   def getByUriAndUserId(uriId: Id[NormalizedURI], userId: Id[User], excludeUriRecommendationState: Option[State[UriRecommendation]])(implicit session: RSession): Option[UriRecommendation] = {
-    (for (row <- rows if row.uriId === uriId && row.userId === userId && row.state =!= excludeUriRecommendationState.orNull) yield row).firstOption
+    (for (row <- byUser(userId)() |> byUri(uriId) if row.state =!= excludeUriRecommendationState.orNull) yield row).firstOption
```
Comment (2is10) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: For predicates whose parameter type leaves no ambiguity, we might consider shortening the method name to `with`:
`(for (row <- with(userId)() |> with(uriId) ...`

(Currently some of the methods begin with `by` and some with `with`. I think I prefer `with`.)

**Comment by 2is10** [comment]
very nice!

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala (hunk 2 above)
Comment (joshm1) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: great idea!

Comment (joshm1) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: actually can't use `with` as a method name, i'll use `by`

File: kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala [hunk 3]
```diff
@@ -43,6 +44,27 @@ class UriRecommendationRepoImpl @Inject() (
 
   type RepoImpl = UriRecommendationTable
 
+  object QueryBuilder {
+    def recommendable(rows: RepoQuery = rows) =
```
Comment (andrewconner) [inline kifi-backend/modules/curator/app/com/keepit/curator/model/UriRecommendationRepo.scala]: I'm not sure these default arguments are a good pattern. It's very easy to forget to call it with the argument, and it "just works", even though in this case you may miss out on a previous clause. @stkem

**Pull request merged**